ENH Refactor Model Card

merveenoyan opened the pull request:
Partially solves #29

merveenoyan commented:
@adrinjalali @BenjaminBossan I cleaned up a little and updated example so you can see the usage.

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -105,4 +99,12 @@
-model_card.save(os.path.join(f"{local_repo}", "README.md"))
+model_card.save(
+    os.path.join(f"{local_repo}", "README.md"),
+    card_data=card_data,
+    limitations=limitations,
+    model_description=model_description,
+    citation_bibtex=citation,
+    model_card_authors=model_card_authors,
+    get_started_code=get_started_code,
```
do we want to have two ways people can set these values? I think it'd be cleaner if the user would only have the `add` method to add these sections (or access through attributes) instead of having them in the `save`'s signature. WDYT @BenjaminBossan ?

merveenoyan replied:
I refactored it accordingly, can you check? it enables method cascading like you asked.

merveenoyan commented:
@adrinjalali we previously talked about how we should enable people to create their own Displays both in #29 and [the model card examination PR](https://github.com/skops-dev/skops/pulls). I looked at [source of Display objects](https://github.com/scikit-learn/scikit-learn/blob/baf0ea25d/sklearn/metrics/_plot/confusion_matrix.py#L12)  and they're essentially matplotlib plots. I say `add_inspection` method doesn't limit to `Display` objects (which are again matplotlib plots) and we take matplotlib plots in general (which I thought of when I was writing the docstring for it). We can put two examples inside the `plot_model_card` one including a matplotlib plot and other including a `Display`.

BenjaminBossan commented:
First of all, thanks for working on this and taking into account the feedback we gave. From my POV, this is going in the right direction.

I don't have a full blown review, but some more high level comments:

In the example, it is still required for the user to create a `CardData` object from `model_cards`. My personal preference would be that we don't require that. Instead, could we not have special fields on the `Card` class like tags, which we would use to populate the `CardData` when the user exports their model card? That way, they don't need to bother with that.

Probably, there is an argument to be made that this card data constitutes meta data and should thus be treated differently. In that case, we could have a special method `add_metadata`. WDYT?

The `save` method may need some cleanup.

- A lot of the code is concerned with determining the template path, I think that should be factored out.
- The method mutates `self.template_sections`, we should avoid this kind of side-effect if possible.
- Do we want to write it in a more generic way? E.g. have an argument like `format='huggingface'` and call a specific method when for each format (only HF at the start).

Regarding the code in `extract_estimator_config`, I would propose that we use [tabulate](https://pypi.org/project/tabulate/) instead of concatenating strings to create tables, which becomes unwieldy very quickly. Of course, we should avoid adding dependencies if possible but tabulate has been very dependable for me in the past and is a very thin library.

merveenoyan commented:
@BenjaminBossan I know about `tabulate` but if this is the only place we tabulate things I'm not in favor of adding an additional dependency. 
I don't see why `template_sections` being mutated is a problem. The saving is where we pass the `template_sections` dictionary so it should be there. 
I can refactor `template_path` related code, sure. Again I'm not nitpicking atm really given it's not open to review yet.
For `CardData` part, sure we can treat it like any other section and let people set things using `add`. 🙂

For this comment
`Do we want to write it in a more generic way? E.g. have an argument like format='huggingface' and call a specific method when for each format (only HF at the start).` I like the idea but I'd love to see what @adrinjalali says.

merveenoyan commented:
I wouldn't like to limit the user from adding as many plots as they wish, does something like below sound ok to you??

```
# append plot_name at the end of the template 
template = open(self.template_sections["template_path"], 'a')
for section in self.template_sections:
    if section not in template:
        template.write("\n\n")
        template.write(f"{section}")
        template.write("{{ " + f"{section}" " }}")
template.close()
```

which simply adds `plot_name` and a slot for it. 
@adrinjalali @BenjaminBossan

BenjaminBossan commented:
> I know about `tabulate` but if this is the only place we tabulate things I'm not in favor of adding an additional dependency.

Okay, let's see if there will be other places we need it. If it's just this single one, I can see your argument.

> I don't see why `template_sections` being mutated is a problem. The saving is where we pass the `template_sections` dictionary so it should be there.

In general, I would try to avoid (non-obvious) side effects if possible, since they make it harder to reason about the code. E.g. it can easily happen that calling the same method twice leads to different outcomes.

> I wouldn't like to limit the user from adding as many plots as they wish

Yes, we shouldn't limit the user. As for the specific implementation, I can't really judge, I would need to better understand the involved code first.

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -7,55 +7,65 @@
+    def __init__(self, model):
+        self.model = model
+        self.hyperparameters_table = self.extract_estimator_config(self.model)
+        self.model_plot = re.sub(r"\n\s+", "", str(estimator_html_repr(model)))
```
a comment on why this `re.sub` needed would help future maintainers.

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -7,55 +7,65 @@
+        self.model_plot = re.sub(r"\n\s+", "", str(estimator_html_repr(model)))
+        self.template_sections = {}
+
+    def add(self, section, value):
+        """Takes values to fill model card template.
+        Parameters:
+        ----------
+        section: str
+                Section in the template.
+        value: str
+                Value to fill in the section.
+        """
+        self.template_sections[section] = value
+        return self
```
I like the API, but I would suggest changing this to:

```suggestion
    def add(self, **kwargs):
        """Takes values to fill model card template.
        Parameters:
        ----------
        **kwargs : dict
            Parameters to be set for the model card. These parameters
            need to be sections of the underlying `jinja` template used.

        """
        for section, value in kwargs.items():
            self.template_sections[section] = value
        return self
```

This would allow users to set multiple values on each call, yet allowing them to also chain the `add` calls as they wish.

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -7,55 +7,65 @@
-
-    return card
+class Card:
+    def __init__(self, model):
```
Are we going to allow them to use an alternative `jinja` template?

If we do, it can be a separate PR though.

BenjaminBossan replied:
Sounds like a good idea, agree that it can be another PR.

osanseviero replied:
From what I see the alternative template can be massed via `card.add(template_path="...")` which I find not super intuitive. I think in the constructor would make more sense. I left a comment at https://github.com/skops-dev/skops/pull/37/files#diff-1ba607fa6851b439c95dfe0e29c812a84e15748682b5c37b049b1fa8c474ba75R113

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -7,55 +7,65 @@
+        )
+        card.save(path)
+
+    def extract_estimator_config(model):
```
We probably want this to be a private method:

```suggestion
    def _extract_estimator_config(model):
```

adrinjalali reviewed (commented):
I think this API looks a lot better. We can move to documentation and tests now IMO.

I left a few comments, but nothing major.

As for `tabulate`, I'd agree with @BenjaminBossan that it'd be useful, but I don't mind not adding it in this PR. I think we'd need more tabulation when we move to adding confusion matrix as a matrix as opposed to a plot and other tabular data we'd like to add in the future. We can add the dependency and refactor when those things happen.

merveenoyan commented:
Model card created with this implementation is [here](https://huggingface.co/merve/sklearn-65b7d701-4875-4baa-9d5d-c9cfa84d66c5) @adrinjalali

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -64,45 +64,50 @@
+plt.savefig(f"{local_repo}/fig2.png")
+
+
+model_card.add_inspection("fig1", "fig1.png")
+model_card.add_inspection("fig2", "fig2.png")
```
should we define a section where these should go? Like, inspection, metrics, etc?

merveenoyan replied:
I actually added a section previously apparently (`self.figure_paths`) but decided not to use it for reasons I don't remember 😅 I addressed this

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,117 @@
+
+        with open(self.template_sections["template_path"], "r") as f:
+            template_content = f.read()
+            breakpoint()
```
probably don't want to have this ;)

merveenoyan replied:
hahahaha sorry 😂

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,117 @@
+        plot_name: str
+                Name of the plot.
+        """
+        self.template_sections[plot_name] = plot_path
```
I would say you can add inspections in a separate attribute than `template_sections` so that bellow in `save` you also don't have to check if the template has a specific name matching this `plot_name` or not.

merveenoyan replied:
You don't necessarily need to match plot, it just looks if template has anything that it used to not have, which is incredibly fragile btw so I'll apply your recommendation😅

merveenoyan replied:
Addressed it.

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,117 @@
+            template_content = f.read()
+            breakpoint()
+            for section in self.template_sections:
+                if section not in template_content and section != "template_path":
```
if you store them in a separate attribute, then this would also not be needed.

merveenoyan replied:
Addressed it.

adrinjalali reviewed (commented):
Looking pretty good, but my previous review still stands, please have a look at them as well :)

merveenoyan commented:
Model card with the latest version: https://huggingface.co/merve/sklearn-9b93776d-fb95-4b81-b267-fa378476bf15

I'll add tests and documentation now @adrinjalali

merveenoyan commented:
@adrinjalali updated the examples, updated tests and added more tests.

merveenoyan commented:
Every time user adds something to template the template file itself changes, maybe it's better if I add a temporary template when user saves the template with added goods.

adrinjalali commented:
You can add your token to the local run by something like: 

```
HF_HUB_TOKEN=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 make doc
```

> Every time user adds something to template the default template file (if it's used) itself changes, maybe it's better if I add a temporary template when user saves the template with added goods.

Yeah the default template shouldn't change.

osanseviero commented:
Is that an actual token? :eyes: I would invalidate it from the Hub if it is

adrinjalali commented:
> Is that an actual token? eyes I would invalidate it from the Hub if it is

That token is already hard coded in `hub_utils/tests/common.py`. Related: #47

merveenoyan commented:
@adrinjalali I added temporary template writing and associated test. With this I think the PR is in a good state. I also fixed docs.

Also all tests pass except for following which are irrelevant? 

```
skops/card/tests/test_card.py::test_save_model_card PASSED
skops/card/tests/test_card.py::test_hyperparameter_table PASSED
skops/card/tests/test_card.py::test_plot_model PASSED
skops/card/tests/test_card.py::test_add PASSED
skops/card/tests/test_card.py::test_add_inspection PASSED
skops/card/tests/test_card.py::test_temporary_plot PASSED
skops/hub_utils/tests/test_hf_hub.py::test_validate_folder PASSED
skops/hub_utils/tests/test_hf_hub.py::test_create_config PASSED
skops/hub_utils/tests/test_hf_hub.py::test_init PASSED
skops/hub_utils/tests/test_hf_hub.py::test_push_download[True] FAILED
skops/hub_utils/tests/test_hf_hub.py::test_push_download[False] FAILED
skops/hub_utils/tests/test_hf_hub.py::test_get_config PASSED
```

merveenoyan commented:
@adrinjalali 
![3v58ic](https://user-images.githubusercontent.com/53175384/180233742-0e01f11e-045c-4ec7-b788-9c2477e41867.jpeg)

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -60,49 +61,64 @@
 # Create a model card
 # ====================
 # We now create a model card, set couple of attributes and save it.
```
This sentence conflicts with the next one, the next one makes it feel like user saves it then adds data

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -60,49 +61,64 @@
+model_card = card.Card(model)
+
+
+# %%
+# Initialize a repository to save our files in
+# ============================================
+# We will now initialize a repository and save our model
```
We should first initialize the local repo, then create the modelcard, which also would be inline with https://github.com/skops-dev/skops/pull/45

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -60,49 +61,64 @@
-    model_card_authors=model_card_authors,
-    get_started_code=get_started_code,
+
+model_card.add(citation="bibtex\n@inproceedings{...,year={2020}}")
+model_card.add(get_started_code=get_started_code).add(
+    model_card_authors=model_card_authors
 )
+model_card.add(limitations=limitations).add(model_description=model_description)
```
```suggestion
model_card.add(
    citation="bibtex\n@inproceedings{...,year={2020}}",
    get_started_code=get_started_code,
    model_card_authors=model_card_authors,
    limitations=limitations,
    model_description=model_description,
)
```

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -60,49 +61,64 @@
+# %%
+# Save model card
+# ===============
+# We can simply save our model card by providing a path to save()
```
```suggestion
# We can simply save our model card by providing a path to ``save()``
```

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -60,49 +61,64 @@
+# ===============
+# We can simply save our model card by providing a path to save()
 
 model_card.save(os.path.join(f"{local_repo}", "README.md"))
```
or:

```suggestion
model_card.save(Path(local_repo) / "README.md")
```

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,120 @@
-    )
-
-    return card
+class Card:
```
needs a class docstring

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,120 @@
+        self.figure_paths = {}
+
+    def add(self, **kwargs):
+        """Takes values to fill model card template.
```
We need to add a `model_card.rst` under `docs/modules` and explain what the available arguments here are, and link to that one from here.

See [here](https://github.com/scikit-learn/scikit-learn/blob/2e49193e3fd9c12a710ecae12f6330c878de0540/sklearn/linear_model/_quantile.py#L29) on how to link to a user guide, and look for `.. _quantile_regression:` in [this file](https://raw.githubusercontent.com/scikit-learn/scikit-learn/main/doc/modules/linear_model.rst) to see an example of a user guide.

If you want to have a separate PR to add the user guide and then add links here, I'm fine with that.

adrinjalali replied:
@merveenoyan has already opened the issue: https://github.com/skops-dev/skops/issues/50, but I think she forgot to put the link here.

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,120 @@
+        """Takes plots to fill model card template.
+        Parameters:
+        ----------
+        Parameters to be set for the model card. These parameters
+        need to be sections of the underlying `jinja` template used.
```
```suggestion
        **kwargs : dict
            The arguments should be of the form `name=plot_path`, where `name`
            is the name of the plot and `plot_path` is the path to the plot,
            relative to the root of the project. The plots should have already
            been saved under the project's folder.
```

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,120 @@
+            self.template_sections[section] = value
+        return self
+
+    def add_inspection(self, **kwargs):
+        """Takes plots to fill model card template.
```
If this adds plots, then we should call it `add_plot`

```suggestion
    def add_plot(self, **kwargs):
        """Add plots to the model card.
        
```

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,120 @@
+        return self
+
+    def save(self, path):
+        """Fills model card template, renders and saves it as markdown file to the target directory.
```
```suggestion
         """Save the model card.

        This method renders the model card in mardown format and then saves it
        as the specified file.
```

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,120 @@
+
+        Parameters:
+        ----------
+        path: Path
```
```suggestion
        path: str, or Path
```

adrinjalali commented on skops/card/tests/test_card.py:
```diff
@@ -17,28 +17,16 @@ def fit_model():
 
-def test_write_model_card():
+def test_save_model_card():
     with tempfile.TemporaryDirectory(prefix="skops-test") as dir_path:
```
You could use a fixture to simplify the tests on the tempdirectory like here:

https://github.com/skops-dev/skops/blob/64366d04a6f49777495b7cce1f13feede8f8b84e/skops/hub_utils/tests/test_hf_hub.py#L25-L28

adrinjalali replied:
This is not fixed yet @merveenoyan .

adrinjalali replied:
you can also have a look at how I'm changing these tests here: https://github.com/skops-dev/skops/pull/45/files#diff-0d678309e10b981c043c9faa2f9b825e9e13fafdce4eed8cc371b8d61874fd26

adrinjalali commented on skops/card/tests/test_card.py:
```diff
@@ -56,4 +44,54 @@ def test_plot_model():
+    # test if the additions are made to a temporary plot
+    # and not to default template or template provided
+    with tempfile.TemporaryDirectory(prefix="skops-test") as dir_path:
+        import skops
```
you can move this to the top of the file

adrinjalali commented on skops/card/tests/test_card.py:
```diff
@@ -56,4 +44,54 @@ def test_plot_model():
+
+
+def test_temporary_plot():
+    # test if the additions are made to a temporary plot
```
```suggestion
    # test if the additions are made to a temporary template file
```

adrinjalali commented on skops/temporary_template.md:
```diff
@@ -0,0 +1,70 @@
+---
```
why is this file here? We should create a temp file on the fly and modify that, and shouldn't have this in the repo.

adrinjalali reviewed (commented):
Really nice, almost there.

BenjaminBossan commented on examples/plot_model_card.py:
```diff
@@ -60,49 +61,64 @@
+    model_card_authors=model_card_authors
 )
+model_card.add(limitations=limitations).add(model_description=model_description)
+predictions = model.predict(X_test)
+cm = confusion_matrix(y_test, predictions, labels=model.classes_)
```
Nit: `y_pred` is more canonical

```suggestion
y_pred = model.predict(X_test)
cm = confusion_matrix(y_test, y_pred, labels=model.classes_)
```

BenjaminBossan commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,120 @@
+        self.model = model
+        self.hyperparameter_table = self._extract_estimator_config()
+        # the spaces in the pipeline breaks markdown, so we replace them
+        self.model_plot = re.sub(r"\n\s+", "", str(estimator_html_repr(model)))
```
Do we want to make that optional? For some models, it doesn't make a lot of sense to include the model plot.

BenjaminBossan replied:
OTOH, I would exclude it if:

- The model is very simple, mainly just a single estimator. The graph provides no additional value since hyper params are already documented.
- The model pipeline is very complex, with dozens of transformers, feature unions etc., making the graph too messy.

BenjaminBossan commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,120 @@
+        card_data_keys = {}
+
+        # if key is supposed to be in metadata and is provided by user, write it to card_data_keys
+        for key in metadata_keys:
+            if key in self.template_sections.keys():
+                card_data_keys[key] = self.template_sections.pop(key, "")
```
I would avoid mutating the `self.template_section`. As an alternative, I would suggest to create a copy at the beginning of the method and perform all operations on that copy:

```python
template_sections = self.template_sections.copy()
...
template_sections.pop(key)
```

Right now, if I call the `save` method twice, the 2nd time around, the `template_sections` will not contain any meta data entries anymore, right?

BenjaminBossan replied:
Nit:
```suggestion
        for key in self.template_sections.keys() & metadata_keys:
            card_data_keys[key] = self.template_sections.pop(key, "")
```

BenjaminBossan commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,120 @@
+            )
+
+        # append plot_name if any plots are provided, at the end of the template
+        if self.figure_paths != {}:
```
```suggestion
        if self.figure_paths:
```

BenjaminBossan commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,120 @@
+                f"{ROOT[0]}/temporary_template.md",
+            )
+            self.template_sections["template_path"] = f"{ROOT[0]}/temporary_template.md"
+            template = open(self.template_sections["template_path"], "a")
```
Let's always use `with open(...)`

BenjaminBossan commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,120 @@
+
+        # append plot_name if any plots are provided, at the end of the template
+        if self.figure_paths != {}:
+            shutil.copyfile(
```
Sorry, I don't quite understand what is going on here. Could you please add a comment to explain why these steps are necessary?

merveenoyan replied:
sure. basically we can't edit jinja template unless we add the slots in the template, that's why I had to append at the end.

BenjaminBossan commented on skops/card/tests/test_card.py:
```diff
@@ -56,4 +44,54 @@ def test_plot_model():
         with open(os.path.join(f"{dir_path}", "README.md"), "r") as f:
             model_card = f.read()
-        assert "<style>" in model_card
+            assert "<style>" in model_card
```
I know this test has already been there but still: Is this really a sufficient test?

merveenoyan replied:
What would you do to assert validity of HTML? I wanted to make this lightweight. 
One think I thought of was to have the content of the HTML somewhere else, load and assert equality.

BenjaminBossan replied:
A style tag could be in the model card for other reasons, I wonder if there is something more specific we could test.

To test such things **super accurately**, we would probably have to parse the document into a tree or something and then query its contents, but I agree that this is not really lightweight.

adrinjalali replied:
I rather not parse the whole thing, but we could look for something more specific, like a parameter name or a class name.

BenjaminBossan commented on skops/card/tests/test_card.py:
```diff
@@ -56,4 +44,54 @@ def test_plot_model():
+
+        ROOT = skops.__path__
+        # read original template
+        f = open(os.path.join(f"{ROOT[0]}", "card", "default_template.md"))
```
Again, let's always use `with open(...)`

BenjaminBossan reviewed (commented):
Good work, thanks Merve.

I have added a few comments, please take a look. Some are just suggestions, feel free to ignore those if you disagree.

merveenoyan commented on skops/card/tests/test_card.py:
```diff
@@ -56,4 +44,54 @@ def test_plot_model():
+        with open(os.path.join(f"{dir_path}", "README.md"), "r") as f:
+            model_card = f.read()
+            assert "![fig1](fig1.png)" in model_card
+        assert os.path.exists(f"{dir_path}/fig1.png")
```
yeah it's tested in another test already. this is just for sanity check 🙂

adrinjalali replied:
I agree with you, but also don't mind it :D

BenjaminBossan replied:
In that case, I'd favor removing it, since it detracts from what is actually being tested.

merveenoyan commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,120 @@
+            self.template_sections[section] = value
+        return self
+
+    def add_inspection(self, **kwargs):
```
It was from either previous PR or the model card issue we were discussing, @adrinjalali picked it.

adrinjalali commented:
There are still outstanding reviews from the previous round @merveenoyan , could you please address those first? I'm happy to do another round of reviews afterwards.

merveenoyan commented:
@adrinjalali I'm going to go through everything and add comment to which ones I solved. Let me know if you need any further changes.

merveenoyan commented:
@adrinjalali added one comment to ask for more clarity + added docs for Card class. other that that, couldn't find anything else that's not addressed. (two of them are left for two different PRs)

osanseviero commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,161 @@
+        card_data.library_name = "sklearn"
+
+        # if template path is not given, use default
+        if self.template_sections.get("template_path") is None:
```
Now that I see this, I'm not sure if the `template_path` should be passed via `add`. I think a more intuitive API would be passing the template in `__init__` as an optional parameter. See suggestion from https://github.com/skops-dev/skops/pull/37#discussion_r925373761

osanseviero replied:
Good idea!

osanseviero commented on examples/plot_model_card.py:
```diff
@@ -56,53 +57,71 @@
-model_card = card.create_model_card(
-    model,
-    card_data=card_data,
+model_card.add(
```
Also given it's very easy to miss a field, should we display some warning is missing some fields from the template? (not sure though, and probably can be handled directly in `modelcards`.

adrinjalali replied:
I think the missing ones are rendered as missing in the `.md` file, so the user would know. I'm not opposed to having a potential warning, but we can probably leave it to another PR. Opened https://github.com/skops-dev/skops/issues/51 to keep track of this.

osanseviero reviewed (commented):
Very very cool work! :fire: I had some fun and uploaded my first `skops`-created repos on the Hub :partying_face: https://huggingface.co/models?library=sklearn&sort=modified

I left a high-level review but this is my first deeper view into the codebase, so please please take things with a grain of salt. Many are questions and a couple are unrelated to the PR, but came up when going through this :) 

Some suggestions would be:
* For discussions in which the agreement is to work in a follow-up PR, I strongly suggest to open an issue to track things
* The utilities are super nice to use once you know how they work, but it took me a while for me to understand how it works. I think documentation for this should be quite extensive, and anything to help users know which arguments are allowed with the default template would be amazing.

I hope my review is useful :)

Great work on this :fire: looking forward to seeing it shipped :rocket:

BenjaminBossan commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,161 @@
+        """
+        for section, value in kwargs.items():
+            self.template_sections[section] = value
+        return self
```
The idea was to enable method chaining ("fluent interface")

adrinjalali reviewed (commented):
The CI also doesn't pass ;)

osanseviero commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,161 @@
+        """Add plots to the model card.
+
+        Parameters:
+        ----------
+        **kwargs : dict
+            The arguments should be of the form `name=plot_path`, where `name`
+            is the name of the plot and `plot_path` is the path to the plot,
+            relative to the root of the project. The plots should have already
+            been saved under the project's folder.
```
Yeah, that is nice!

BenjaminBossan replied:
> But it's also putting a constraint on plot name, i.e. it has to be a valid identifier

Not necessarily, you can do:

```python
model.add_plot(**{"not a valid identifier": foo})
```

It still has to be a string but that should be okay.

adrinjalali replied:
Nice, then we don't need to change anything. I think it's a nice syntax.

osanseviero commented on examples/plot_model_card.py:
```diff
@@ -56,53 +57,71 @@
+local_repo = mkdtemp(prefix="skops-")
+
+hub_utils.init(
+    model=pkl_name, requirements=[f"scikit-learn={sklearn.__version__}"], dst=local_repo
```
I see! Thanks for explaining!

merveenoyan commented:
doctest expects nothing but `model_card.add()` line returns self so I'll either do something like
```
    >>> model_card.add(license="mit")
    <skops.card._model_card.Card object at 0x7fd35cf2bcd0>
```
or
`    >>> model_card = model_card.add(license="mit")`
which I don't know what's worse so I picked latter, but user needs to know it's not how they should use this? 😅😂

adrinjalali commented:
If you don't want the user to care about the output, you can do:

```python
>>> model_card.add(license="mit")
...
```

The ellipsis (`...`) will match everything.

We might need to enable it though, but it should be enabled by default I think.

adrinjalali commented:
Other than that, I think it's best if we actually implement a basic `__repr__` and `__str__` for the `Card` class.

adrinjalali commented:
can you give me the short hash of the commit where it fails?

merveenoyan commented:
In CI, when you run doctest you use
`python -m pytest -s --cov=skops --doctest-modules skops/`
meanwhile it should be something like
`python -m doctest -o=ELLIPSIS ` otherwise it doesn't get that it should be by default like that.

merveenoyan commented:
WHY though
```
040     >>> model_card.add(license="mit") # doctest: +ELLIPSIS
Expected nothing
Got:
    <skops.card._model_card.Card object at 0x7fbba93d2c10>
```

what am I doing wrong
```
>>> model_card.add(license="mit") # doctest: +ELLIPSIS
    ...
```
I've read around 98392 stackoverflow answers and doctest docs but I can't make this work.

merveenoyan commented:
After fighting doctest and winning
![image](https://user-images.githubusercontent.com/53175384/180840745-9f6670da-e735-4a19-ad88-74c39b0bf250.png)

merveenoyan commented:
![eb834d4d-824d-4d74-a4b5-1913cfad4403_text](https://user-images.githubusercontent.com/53175384/180840971-af66478b-56e4-491d-a945-f7118968199a.gif)

adrinjalali commented on examples/plot_hf_hub.py:
```diff
@@ -86,8 +85,8 @@
 # %%
 # Model Card
 # ==========
-card_data = CardData(tags=["tabular-classification"])
```
The good thing about sphinx-gallery is that it actually runs these scripts and creates a notebook and an html file for them, for this CI, it's rendered here: https://skops--37.org.readthedocs.build/en/37/auto_examples/plot_model_card.html#sphx-glr-auto-examples-plot-model-card-py

That means the CI would only become green if the script runs ;)

osanseviero replied:
This is super cool, thanks for sharing!

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,189 @@
+    ... display_labels=model.classes_)
+    >>> disp.plot() # doctest: +ELLIPSIS
+    <sklearn.metrics._plot.confusion_matrix.ConfusionMatrixDisplay object at ...>
+    >>> plt.savefig("confusion_matrix.png") # doctest: +ELLIPSIS
+    ...
+    >>> model_card.add_plot(confusion_matrix="confusion_matrix.png") # doctest: +ELLIPSIS
+    <skops.card._model_card.Card object at ...>
+    >>> model_card.save((Path("save_dir") / "README.md")) # doctest: +ELLIPSIS
+    ...
```
We can enable ELLIPSIS everywhere in our config file. Opened https://github.com/skops-dev/skops/issues/58 to track this. Nothing to do here.

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,189 @@
+
+        # if key is supposed to be in metadata and is provided by user, write it to card_data_keys
+        for key in self.template_sections.keys() & metadata_keys:
+            card_data_keys[key] = self.template_sections.pop(key, "")
```
this line is not tested

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -56,53 +57,71 @@
+disp.plot()
 
-_, pkl_name = mkstemp(prefix="skops-", suffix=".pkl")
+plt.savefig(f"{local_repo}/confusion_matrix.png")
```
instead of using `plt` directly, better to use the `figure_` attribute:

``` python
disp.figure.savefig(...)
```

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -56,53 +57,71 @@
 
-with open(pkl_name, mode="bw") as f:
-    pickle.dump(model, file=f)
+model_card.add_plot(confusion_matrix="confusion_matrix.png")
```
to show case how users can have an arbitrary string as plot name:

``` python
 model_card.add_plot(**{"confusion matrix": "confusion_matrix.png"})
```

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,189 @@
+        ----------
+        **kwargs : dict
+            Parameters to be set for the model card. These parameters
+            need to be sections of the underlying `jinja` template used.
```
need a `Returns` section for `self`.

adrinjalali commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,189 @@
+            is the name of the plot and `plot_path` is the path to the plot,
+            relative to the root of the project. The plots should have already
+            been saved under the project's folder.
+        """
```
same for this docstring

adrinjalali reviewed (commented):
Some final nits, then I'm happy to merge.

merveenoyan commented:
@adrinjalali codecov test on windows fails for some reason. it wasn't failing before.

adrinjalali commented:
@merveenoyan seems like an issue on codecov side, re-running or waiting a few hours seems to be the suggested action lol

adrinjalali commented on examples/plot_model_card.py:
```diff
@@ -56,53 +56,71 @@
+disp.plot()
 
-_, pkl_name = mkstemp(prefix="skops-", suffix=".pkl")
+disp.figure_.savefig((Path(local_repo) / "confusion_matrix.png"))
```
I don't think you need those extra parenthesis

```suggestion
disp.figure_.savefig(Path(local_repo) / "confusion_matrix.png")
```

adrinjalali reviewed (approved):
LGTM!

adrinjalali commented:
Since this is quite significant, I rather not merge with one approval.

BenjaminBossan commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,195 @@
+            self._figure_paths[plot_name] = plot_path
+        return self
+
+    def save(self, path):
```
I have a few issues with this function, please let me know if you agree:

1. It mutates `self.template_sections`

As a user, I would be surprised to find that the `save` method actually mutates the object itself. At the moment, that might not cause a bug, but this type of hidden mutation has the tendency to come back to bite you. Therefore, I would avoid it except if there is no good reason.

To achieve this, simply declare a copy at the beginning of the method, such as `template_sections = self.template_sections.copy()`. After that, each time you use `self.template_sections` in this method, use `template_sections` instead.

It would be even safer to use `copy.deepcopy` but here we're creating a whole duplicate, I'm not sure if that can be too expensive in some circumstances (doesn't seem so but I'm not sure).

If you agree with this change, here is a unit test that would fail without the change but passes with the change:

```python
import copy
...
def test_template_sections_not_mutated_by_save(destination_path, model_card):
    # TODO
    template_sections_before = copy.deepcopy(model_card.template_sections)
    model_card.save(Path(destination_path) / "README.md")
    template_sections_after = copy.deepcopy(model_card.template_sections)
    assert template_sections_before == template_sections_after
```

2. Remove `if self._figure_paths:` branch

I saw that there is some code duplication that we can avoid. It is not strictly necessary to check `if self._figure_paths` and then do an `else`. The reason it can be avoided is because the `for plot in self._figure_paths` already is a no-op if `self._figure_paths` is empty, so the outcome is the same. Thus this whole code:

```python
        if self._figure_paths:
            with tempfile.TemporaryDirectory() as tmpdirname:
                shutil.copyfile(
                    self.template_sections["template_path"],
                    f"{tmpdirname}/temporary_template.md",
                )
                #  create a temporary template with the additional plots
                self.template_sections[
                    "template_path"
                ] = f"{tmpdirname}/temporary_template.md"
                # add plots at the end of the template
                with open(self.template_sections["template_path"], "a") as template:
                    for plot in self._figure_paths:
                        template.write(
                            f"\n\n{plot}\n"
                            + f"![{plot}]({self._figure_paths[plot]})\n\n"
                        )
                card = ModelCard.from_template(
                    card_data=card_data,
                    hyperparameter_table=self.hyperparameter_table,
                    model_plot=self.model_plot,
                    **self.template_sections,
                )

        else:
            card = ModelCard.from_template(
                card_data=card_data,
                hyperparameter_table=self.hyperparameter_table,
                model_plot=self.model_plot,
                **self.template_sections,
            )
```

can be simplified to:

```python
        with tempfile.TemporaryDirectory() as tmpdirname:
            shutil.copyfile(
                template_sections["template_path"],
                f"{tmpdirname}/temporary_template.md",
            )
            #  create a temporary template with the additional plots
            template_sections["template_path"] = f"{tmpdirname}/temporary_template.md"
            # add plots at the end of the template
            with open(template_sections["template_path"], "a") as template:
                for plot in self._figure_paths:
                    template.write(
                        f"\n\n{plot}\n"
                        + f"![{plot}]({self._figure_paths[plot]})\n\n"
                    )

            card = ModelCard.from_template(
                card_data=card_data,
                hyperparameter_table=self.hyperparameter_table,
                model_plot=self.model_plot,
                **template_sections,
            )
```

There is, however, a small disadvantage to the proposed change: It will result in the temporary copy of the template file to be made even if there is no plot. IMHO this is a small price to pay for the code simplification but feel free to disagree.

3. Breaking down the method

This should probably be done in another PR but I wanted to write it down or else I'll forget:

The `save` method does too many different things at the same time and should be broken down. E.g. there could be one sub-method to get the `card_data` and another one to get the `card`.

BenjaminBossan reviewed (commented):
We're almost good to go.

The only real issue I had was still with the `save` method implementation. Please take a look at my suggestions.

merveenoyan commented:
@BenjaminBossan addressed your comments, used deepcopy.

BenjaminBossan commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,187 @@
+        card_data_keys = {}
+
+        # if key is supposed to be in metadata and is provided by user, write it to card_data_keys
+        for key in self.template_sections.keys() & metadata_keys:
+            card_data_keys[key] = self.template_sections.pop(key, "")
```
```suggestion
        for key in template_sections.keys() & metadata_keys:
            card_data_keys[key] = template_sections.pop(key, "")
```

Should also use `template_sections`, not `self.template_sections`, right?

BenjaminBossan commented on skops/card/_model_card.py:
```diff
@@ -1,61 +1,187 @@
+
+        # copying the template so that the original template is not touched/changed
+        # append plot_name if any plots are provided, at the end of the template
+        # if any plot is given, copy the template to a different path
```
```suggestion
```
IIUC, the comment is not necessary anymore after the change.

BenjaminBossan reviewed (approved):
Thanks, a big change, happy to see it ready.
